sqlalchemy-stubs/util/_collections.pyi

```diff
@@ -1,101 +1,142 @@
 
-class OrderedDict(dict):
+class OrderedDict(Dict[_KT, _VT]):
     def __reduce__(self): ...
     def __init__(self, ____sequence: Optional[Any] = ..., **kwargs) -> None: ...
-    def clear(self): ...
-    def copy(self): ...
-    def __copy__(self): ...
-    def sort(self, *arg, **kw): ...
-    def update(self, ____sequence: Optional[Any] = ..., **kwargs): ...
-    def setdefault(self, key, value): ...
-    def __iter__(self): ...
-    def keys(self): ...
-    def values(self): ...
-    def items(self): ...
-    def itervalues(self): ...
-    def iterkeys(self): ...
-    def iteritems(self): ...
-    def __setitem__(self, key, object): ...
-    def __delitem__(self, key): ...
-    def pop(self, key, *default): ...
-    def popitem(self): ...
-
-class OrderedSet(set):
-    def __init__(self, d: Optional[Any] = ...) -> None: ...
-    def add(self, element): ...
-    def remove(self, element): ...
-    def insert(self, pos, element): ...
-    def discard(self, element): ...
-    def clear(self): ...
-    def __getitem__(self, key): ...
-    def __iter__(self): ...
-    def __add__(self, other): ...
-    def update(self, iterable): ...
-    __ior__: Any = ...
-    def union(self, other): ...
-    __or__: Any = ...
-    def intersection(self, other): ...
-    __and__: Any = ...
-    def symmetric_difference(self, other): ...
-    __xor__: Any = ...
-    def difference(self, other): ...
-    __sub__: Any = ...
-    def intersection_update(self, other): ...
-    __iand__: Any = ...
-    def symmetric_difference_update(self, other): ...
-    __ixor__: Any = ...
-    def difference_update(self, other): ...
-    __isub__: Any = ...
+    def clear(self) -> None: ...
+    def copy(self) -> OrderedDict[_KT, _VT]: ...
+    def __copy__(self) -> OrderedDict[_KT, _VT]: ...
+    if sys.version_info < (3, 0):
+        def sort(self, cmp: Callable[[_VT, _VT], Any] = ..., key: Callable[[_VT], Any] = ..., reverse: bool = ...) -> None: ...
+    else:
+        def sort(self, *, key: Optional[Callable[[_VT], Any]] = ..., reverse: bool = ...) -> None: ...
+    @overload
+    def update(self, ____sequence: Mapping[_KT, _VT], **kwargs: _VT) -> None: ...
+    @overload
+    def update(self, ____sequence: Iterable[Tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
+    @overload
+    def update(self, **kwargs: _VT) -> None: ...
+    # Signature of "setdefault" incompatible with supertype "dict" and "MutableMapping"
+    def setdefault(self, key: _KT, value: _VT) -> _VT: ...  # type: ignore
+    def __iter__(self) -> Iterator[_KT]: ...
+    # Return type of "keys" incompatible with supertype "dict" and "Mapping"
+    def keys(self) -> List[_KT]: ...  # type: ignore
+    # Return type of "values" incompatible with supertype "dict" and "Mapping"
+    def values(self) -> List[_VT]: ...  # type: ignore
+    # Return type of "items" incompatible with supertype "dict" and "Mapping"
+    def items(self) -> List[Tuple[_KT, _VT]]: ...  # type: ignore
```

Comment:
It is a bit sad to have such a bunch of type ignores. Are these here just because the superclass returns an `Iterator`? Also are you sure this is still true on Python 3? If yes, please add a comment. Otherwise, maybe you can also use `if sys.version_info...` here? But be careful, mypy doesn't recognize all forms. I would add version specific tests for these.

Reply:
Take `OrderedDict.items`. The code for it is:

```python
    def items(self):
        return [(key, self[key]) for key in self._list]
```

The definition for `builtins.dict` is:

```python
class dict(MutableMapping[_KT, _VT], Generic[_KT, _VT]):
    # ...
    def items(self) -> ItemsView[_KT, _VT]: ...
```

and the definition for `typing.Mapping` is:

```python
class Mapping(_Collection[_KT], Generic[_KT, _VT_co]):
    # ...
    def items(self) -> AbstractSet[Tuple[_KT, _VT_co]]: ...
```

A lot of these ignores have similar situations.